fem/particleset.hpp

```diff
@@ -0,0 +1,521 @@
+{
+private:
+   /// Constructs an Array of size N filled with Ordering::Type o.
+   static Array<Ordering::Type> GetOrderingArray(Ordering::Type o, int N);
+   /// Returns default field name for field index i. "Field_{i}"
+   static std::string GetDefaultFieldName(int i);
+   /// Returns default tag name for tag index i. "Tag_{i}"
+   static std::string GetDefaultTagName(int i);
+   /// Constructs an Array of size N filled with nullptr.
+   static Array<const char*> GetEmptyNameArray(int N);
+#ifdef MFEM_USE_MPI
+   static unsigned int GetRank(MPI_Comm comm_);
+   static unsigned int GetSize(MPI_Comm comm_);
+#endif // MFEM_USE_MPI
+
+protected:
+
+   /// Stride for IDs (when new particles are added).
+   /// In parallel, this defaults to the number of MPI ranks.
+   const unsigned int id_stride;
+
+   /// Current ID to be assigned to the next particle added.
+   /// In parallel, this starts locally as the rank and increments with
+   /// id_stride, ensuring a global unique identifier whenever a particle is
+   /// added.
+   unsigned int id_counter;
+
+   /// Array holding all particles' globally unique ID.
+   Array<unsigned int> ids;
+
+   /// All particle coordinates.
+   ParticleVector coords;
+
+   /// All particle fields.
+   std::vector<std::unique_ptr<ParticleVector>> fields;
+
+   /// All particle tags.
+   std::vector<std::unique_ptr<Array<int>>> tags;
+
+   /// Field names, to be written when \ref PrintCSV is called.
+   std::vector<std::string> field_names;
+
+   /// Tag names, to be written when \ref PrintCSV is called.
+   std::vector<std::string> tag_names;
+
+   /// Add particles with global identifiers \p new_ids and optionally get the
+   /// local indices of new particles in \p new_indices .
+   /// Note the data of new particles is uninitialized and must be set
+   /// using \ref SetParticle
+   void AddParticles(const Array<unsigned int> &new_ids,
+                     Array<int> *new_indices=nullptr);
+
+
+#ifdef MFEM_USE_MPI
+   MPI_Comm comm;
+#endif // MFEM_USE_MPI
+
+
+#if defined(MFEM_USE_MPI) && defined(MFEM_USE_GSLIB)
+   std::unique_ptr<gslib::comm> gsl_comm;
+   std::unique_ptr<gslib::crystal> cr;
+
+   /// \cond DO_NOT_DOCUMENT
+   template<std::size_t NBytes>
+   static void TransferParticlesImpl(ParticleSet &pset,
+                                     const Array<unsigned int> &send_idxs, const Array<unsigned int> &send_ranks);
+
+   using TransferParticlesType = void (*)(ParticleSet &pset,
+                                          const Array<unsigned int> &send_idxs,
+                                          const Array<unsigned int> &send_ranks);
+
+   // Specialization parameter: NBytes
+   MFEM_REGISTER_KERNELS(TransferParticles, TransferParticlesType, (int));
+   friend TransferParticles;
+   struct Kernels
+   {
+      Kernels();
+   };
+   /// \endcond DO_NOT_DOCUMENT
+
+#endif // MFEM_USE_MPI && MFEM_USE_GSLIB
+
+   /// Create a \ref Particle object with the same spatial dimension, number of
```

Comment:
Again, rather drop that `\ref` as it just decreases readability and it is linked anyway.